File: day03/star06.py
def score(c):
    s = 1 + ord(c)
    s -= ord('a') if c.islower() else (ord('A') - 26)
    return s

def find_badge(lines):
    s1, s2, s3 = map(set, lines)
    badge = (s1 & s2 & s3).pop()
    return badge

with open('input.txt') as f:
    total = 0
    while True:
        lines = [f.readline().strip() for _ in range(3)]
        if not lines[0]:
            break
        total += score(find_badge(lines))

print(f"{total = }")

File: day03/star06.cpp
#include <fstream>
#include <string>
#include <iostream>
#include <cctype>
#include <array>

int score(char c) {
    return std::islower(c) ? 1 + c - 'a' : 27 + c - 'A';
}

char find_badge(std::array<std::string, 3> lines) {
    for (auto c : lines[0]) {
        if (lines[1].find(c) != std::string::npos && lines[2].find(c) != std::string::npos) {
            return c;
        }
    }
    return '\0';
}

int main() {
    std::ifstream f {"input.txt"};
    std::string line, left, right;
    std::array<std::string, 3> lines;
    
    int total {0};
    int i {0};
    while (std::getline(f, line)) {
        lines[i++] = line;
        if (i == 3) {
            i = 0;
            total += score(find_badge(lines));
        }
    }

    std::cout << "total = " << total << std::endl;
}